File: app/test/poe_flip_finder/divination_card_opportunity_finder_test.exs
defmodule PoeFlipFinder.DivinationCardOpportunityFinderTest do
  use ExUnit.Case, async: true

  alias PoeFlipFinder.{
    BaseCurrencyIds,
    Currency,
    DivinationCardOpportunityFinder,
    DivinationCardReward,
    DivineChaosRate,
    ExchangeMarketSnapshot,
    League
  }

  # One test per use case enumerated in docs/PRD.md § 7.3, per
  # docs/ELIXIR_TEST_MANIFESTO.md's Use-Case Discovery Procedure. Ported
  # from the deleted Java DivinationCardOpportunityFinderTest -- see the
  # elixir-migration merge commit history for the original.

  @now ~U[2026-08-06 12:00:00Z]
  @league %League{
    id: 1,
    external_id: "Standard",
    display_name: "Standard",
    is_current: false,
    has_exchange_activity: true
  }
  @chaos %Currency{
    id: 1,
    external_id: BaseCurrencyIds.chaos_external_id(),
    display_name: "Chaos Orb",
    item_type: :currency
  }
  @divine %Currency{
    id: 2,
    external_id: BaseCurrencyIds.divine_external_id(),
    display_name: "Divine Orb",
    item_type: :currency
  }
  @rate_210 %DivineChaosRate{
    chaos_currency: @chaos,
    divine_currency: @divine,
    chaos_per_divine: 210.0
  }

  defp reward(overrides) do
    defaults = %{
      card: %Currency{external_id: nil, display_name: "Test Card", item_type: :divination_card},
      stack_size: 8,
      reward_currency: %Currency{external_id: nil, display_name: "Test Reward", item_type: :currency},
      reward_quantity: 5,
      predictable: true
    }

    struct(DivinationCardReward, Map.merge(defaults, overrides))
  end

  defp snapshot(currency_a, currency_b, opts \\ []) do
    ratios = Keyword.get(opts, :ratios, {1, 5, 1, 5})
    {ra, rb, ra2, rb2} = ratios
    stocks = Keyword.get(opts, :stocks, {100, 100, 1, 1})
    {sa_lo, sa_hi, sb_lo, sb_hi} = stocks

    %ExchangeMarketSnapshot{
      id: 1,
      generation_id: 999,
      league: @league,
      currency_a: currency_a,
      currency_b: currency_b,
      snapshot_hour: @now,
      volume_traded_a: 100,
      volume_traded_b: 100,
      lowest_stock_a: sa_lo,
      highest_stock_a: sa_hi,
      lowest_stock_b: sb_lo,
      highest_stock_b: sb_hi,
      lowest_ratio_a: ra,
      highest_ratio_a: ra2,
      lowest_ratio_b: rb,
      highest_ratio_b: rb2
    }
  end

  defp card_currency, do: %Currency{external_id: "Metadata/Items/DivinationCards/DivinationCardTest", display_name: "Test Card", item_type: :divination_card}
  defp reward_currency, do: %Currency{external_id: "Metadata/Items/Currency/CurrencyTestReward", display_name: "Test Reward", item_type: :currency}

  test "golden path: hand-verified buy + resale via chaos" do
    # Buy leg: flat 1:5 chaos:card ratio -> suggested_buy_price=4 (card per
    # chaos), buy_leg_stock=100. cost_in_base = stack_size(8)/4 = 2 chaos.
    buy_snapshot = snapshot(@chaos, card_currency(), ratios: {1, 5, 1, 5})

    # Resale leg: flat 1:2 chaos:reward ratio -> market_sell_price=2 (reward
    # per chaos). resale = reward_quantity(5)/2 = 2.5 chaos.
    resale_snapshot = snapshot(@chaos, reward_currency(), ratios: {1, 2, 1, 2})

    [opportunity] =
      DivinationCardOpportunityFinder.find([buy_snapshot, resale_snapshot], [reward(%{})], @rate_210)

    assert opportunity.technique == :divination_card
    assert opportunity.start == [%PoeFlipFinder.CurrencyAmount{currency: @chaos, quantity: 2.0}]
    assert opportunity.sell == [%PoeFlipFinder.CurrencyAmount{currency: @chaos, quantity: 2.5}]
    assert opportunity.profit.quantity == 0.5
    assert opportunity.margin_percent == 25.0
    assert opportunity.volume == 100
    assert opportunity.detail == "≈0.25c per card"
  end

  test "predictable: false cards are always excluded" do
    buy_snapshot = snapshot(@chaos, card_currency())
    resale_snapshot = snapshot(@chaos, reward_currency(), ratios: {1, 2, 1, 2})

    assert DivinationCardOpportunityFinder.find(
             [buy_snapshot, resale_snapshot],
             [reward(%{predictable: false})],
             @rate_210
           ) == []
  end

  test "no chaos/divine rate: contributes nothing" do
    buy_snapshot = snapshot(@chaos, card_currency())
    resale_snapshot = snapshot(@chaos, reward_currency(), ratios: {1, 2, 1, 2})

    assert DivinationCardOpportunityFinder.find([buy_snapshot, resale_snapshot], [reward(%{})], nil) == []
  end

  test "buy leg falls back to Divine when no Chaos market exists for the card" do
    buy_snapshot = snapshot(@divine, card_currency())
    resale_snapshot = snapshot(@chaos, reward_currency(), ratios: {1, 2, 1, 2})

    [opportunity] =
      DivinationCardOpportunityFinder.find([buy_snapshot, resale_snapshot], [reward(%{})], @rate_210)

    assert [%PoeFlipFinder.CurrencyAmount{currency: @divine}] = opportunity.start
  end

  test "missing buy leg (no market in either Chaos or Divine): card silently dropped" do
    resale_snapshot = snapshot(@chaos, reward_currency(), ratios: {1, 2, 1, 2})

    assert DivinationCardOpportunityFinder.find([resale_snapshot], [reward(%{})], @rate_210) == []
  end

  test "resale leg: reward currency is Chaos itself needs no market lookup" do
    buy_snapshot = snapshot(@chaos, card_currency())

    [opportunity] =
      DivinationCardOpportunityFinder.find(
        [buy_snapshot],
        [reward(%{reward_currency: %Currency{external_id: nil, display_name: "Chaos Orb", item_type: :currency}, reward_quantity: 7})],
        @rate_210
      )

    assert opportunity.sell == [%PoeFlipFinder.CurrencyAmount{currency: @chaos, quantity: 7}]
  end

  test "resale leg: reward currency is Divine itself converts via chaos_per_divine" do
    buy_snapshot = snapshot(@chaos, card_currency())

    [opportunity] =
      DivinationCardOpportunityFinder.find(
        [buy_snapshot],
        [reward(%{reward_currency: %Currency{external_id: nil, display_name: "Divine Orb", item_type: :currency}, reward_quantity: 2})],
        @rate_210
      )

    assert opportunity.sell == [%PoeFlipFinder.CurrencyAmount{currency: @chaos, quantity: 420.0}]
  end

  test "resale leg falls back to Divine market when no Chaos market exists for the reward" do
    buy_snapshot = snapshot(@chaos, card_currency())
    # Only a Divine-side market for the reward exists -- flat 1:2 ->
    # market_sell_price=2 (reward per divine). resale = 4/2 * 210 = 420 chaos.
    resale_snapshot = snapshot(@divine, reward_currency(), ratios: {1, 2, 1, 2})

    [opportunity] =
      DivinationCardOpportunityFinder.find(
        [buy_snapshot, resale_snapshot],
        [reward(%{reward_quantity: 4})],
        @rate_210
      )

    assert opportunity.sell == [%PoeFlipFinder.CurrencyAmount{currency: @chaos, quantity: 420.0}]
  end

  test "missing resale leg (reward has no market against Chaos or Divine): card silently dropped" do
    buy_snapshot = snapshot(@chaos, card_currency())

    assert DivinationCardOpportunityFinder.find([buy_snapshot], [reward(%{})], @rate_210) == []
  end

  test "buy leg re-orients when the card is worth more than 1 Chaos (the Sephirot case)" do
    # Direct orientation (card units per 1 chaos): ratio 200:1 -> price
    # 1/200 = 0.005, floors to 0, suggested_buy_price goes negative -> nil,
    # not viable. Without the fallback this card would be silently dropped
    # despite having a real, tradeable market.
    expensive_card = snapshot(@chaos, card_currency(), ratios: {200, 1, 200, 1}, stocks: {1, 1, 20, 30})
    resale_snapshot = snapshot(@chaos, reward_currency(), ratios: {1, 2, 1, 2})

    [opportunity] =
      DivinationCardOpportunityFinder.find(
        [expensive_card, resale_snapshot],
        [reward(%{stack_size: 1})],
        @rate_210
      )

    # Inverted orientation (chaos per 1 card): price 200 -> suggested_buy_price
    # = floor(200) - 1 = 199. cost = stack_size(1) * 199 = 199 chaos.
    assert opportunity.start == [%PoeFlipFinder.CurrencyAmount{currency: @chaos, quantity: 199.0}]
    assert opportunity.volume == 30
  end

  test "resale leg re-orients when the reward is worth more than 1 Chaos" do
    buy_snapshot = snapshot(@chaos, card_currency())
    # Direct orientation (reward per 1 chaos): ratio 200:1 -> price 0.005,
    # market_sell_price floors to 0, not viable (0 is not > 0). Inverted
    # orientation (chaos per 1 reward unit): price 200, market_sell_price=200.
    expensive_reward = snapshot(@chaos, reward_currency(), ratios: {200, 1, 200, 1})

    [opportunity] =
      DivinationCardOpportunityFinder.find(
        [buy_snapshot, expensive_reward],
        [reward(%{reward_quantity: 1})],
        @rate_210
      )

    assert opportunity.sell == [%PoeFlipFinder.CurrencyAmount{currency: @chaos, quantity: 200.0}]
  end

  test "resale is priced at the market rate, never the round-trip-favorable rate" do
    buy_snapshot = snapshot(@chaos, card_currency())
    # Same shape as the hand-verified Bulk Buy chaos-wisdom fixture:
    # price_at_lowest=185, price_at_highest=366 -> market_sell_price=366,
    # but suggested_sell_price (round-trip) would be floor(185)+1=186.
    # reward_quantity=366 makes the two divisors trivially distinguishable:
    # 366/366=1.0 if market_sell_price is used (correct), 366/186!=1.0 if
    # suggested_sell_price were wrongly used instead.
    resale_snapshot = snapshot(@chaos, reward_currency(), ratios: {1, 185, 1, 366})

    [opportunity] =
      DivinationCardOpportunityFinder.find(
        [buy_snapshot, resale_snapshot],
        [reward(%{reward_quantity: 366})],
        @rate_210
      )

    assert opportunity.sell == [%PoeFlipFinder.CurrencyAmount{currency: @chaos, quantity: 1.0}]
  end
end

File: backend/src/test/java/com/poeflipfinder/backend/usecase/computeflipopportunities/DivinationCardOpportunityFinderTest.java
package com.poeflipfinder.backend.usecase.computeflipopportunities;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.within;

import com.poeflipfinder.backend.entity.Currency;
import com.poeflipfinder.backend.entity.DivinationCardReward;
import com.poeflipfinder.backend.entity.ExchangeMarketSnapshot;
import com.poeflipfinder.backend.entity.FlipOpportunity;
import com.poeflipfinder.backend.entity.League;
import com.poeflipfinder.backend.entity.Technique;
import java.time.Instant;
import java.util.List;
import org.junit.jupiter.api.Test;

class DivinationCardOpportunityFinderTest {

    private static final Instant NOW = Instant.parse("2026-08-06T12:00:00Z");

    private final League league = new League(1L, "Standard", "Standard", false, true);
    private final Currency chaos =
            new Currency(1L, "Metadata/Items/Currency/CurrencyRerollRare", "Chaos Orb", null, Currency.ItemType.CURRENCY);
    private final Currency divine =
            new Currency(2L, "Metadata/Items/Currency/CurrencyModValues", "Divine Orb", null, Currency.ItemType.CURRENCY);
    private final Currency card = new Currency(
            3L, "Metadata/Items/DivinationCards/DivinationCardStackedDeck", "Stacked Deck", null,
            Currency.ItemType.DIVINATION_CARD);
    private final Currency regal =
            new Currency(4L, "Metadata/Items/Currency/CurrencyUpgradeMagicToRare", "Regal Orb", null, Currency.ItemType.CURRENCY);

    private final DivinationCardOpportunityFinder finder = new DivinationCardOpportunityFinder();
    private final DivineChaosRate rate210 = new DivineChaosRate(chaos, divine, 210.0);

    // Name-only placeholder Currency, matching what BundledDivinationCardReferenceGateway produces.
    private Currency placeholder(String name, Currency.ItemType itemType) {
        return new Currency(null, null, name, null, itemType);
    }

    private DivinationCardReward reward(int stackSize, String rewardCurrencyName, int rewardQuantity) {
        return new DivinationCardReward(
                placeholder("Stacked Deck", Currency.ItemType.DIVINATION_CARD),
                stackSize,
                placeholder(rewardCurrencyName, Currency.ItemType.CURRENCY),
                rewardQuantity,
                true);
    }

    private ExchangeMarketSnapshot snapshot(
            Currency currencyA,
            Currency currencyB,
            double lowestRatioA,
            double lowestRatioB,
            double highestRatioA,
            double highestRatioB,
            long lowestStockA,
            long highestStockA,
            long lowestStockB,
            long highestStockB) {
        return new ExchangeMarketSnapshot(
                1L, 999L, league, currencyA, currencyB, NOW, 100, 100,
                lowestStockA, highestStockA, lowestStockB, highestStockB,
                lowestRatioA, highestRatioA, lowestRatioB, highestRatioB);
    }

    @Test
    void find_chaosRewardThroughChaosMarket_producesHandVerifiedOpportunity() {
        // Chaos-card leg: priceAtLowest=8, priceAtHighest=13 -> buy=floor(13)-1=12, buyLegStock=highestStockA=50.
        ExchangeMarketSnapshot chaosLeg = snapshot(chaos, card, 1, 8, 1, 13, 100, 50, 1, 1);
        DivinationCardReward stackedDeckReward = reward(8, "Chaos Orb", 5);

        List<FlipOpportunity> opportunities = finder.find(List.of(chaosLeg), List.of(stackedDeckReward), rate210);

        assertThat(opportunities).hasSize(1);
        FlipOpportunity opportunity = opportunities.get(0);
        assertThat(opportunity.technique()).isEqualTo(Technique.DIVINATION_CARD);
        assertThat(opportunity.start().get(0).currency()).isEqualTo(chaos);
        assertThat(opportunity.start().get(0).quantity()).isCloseTo(8.0 / 12, within(1e-6));
        assertThat(opportunity.via()).hasSize(2);
        assertThat(opportunity.via().get(0).currency()).isEqualTo(card);
        assertThat(opportunity.via().get(0).quantity()).isEqualTo(8.0);
        assertThat(opportunity.via().get(1).currency()).isEqualTo(chaos);
        assertThat(opportunity.via().get(1).quantity()).isEqualTo(5.0);
        assertThat(opportunity.sell().get(0).currency()).isEqualTo(chaos);
        assertThat(opportunity.sell().get(0).quantity()).isEqualTo(5.0);
        double costChaos = 8.0 / 12;
        assertThat(opportunity.marginPercent()).isCloseTo((5.0 - costChaos) / costChaos * 100, within(0.01));
        assertThat(opportunity.profit().currency()).isEqualTo(chaos);
        assertThat(opportunity.profit().quantity()).isCloseTo(5.0 - costChaos, within(1e-6));
        assertThat(opportunity.volume()).isEqualTo(50);
        assertThat(opportunity.detail()).isEqualTo("≈0.08c per card");
    }

    @Test
    void find_rewardInAThirdCurrency_pricedAtMarketRate_notCompetitiveRate() {
        // Same chaos-card buy leg as above (suggestedBuyPrice=12).
        ExchangeMarketSnapshot chaosCardLeg = snapshot(chaos, card, 1, 8, 1, 13, 100, 50, 1, 1);
        // Chaos-regal leg: priceAtLowest=2, priceAtHighest=3 -> marketSellPrice=floor(3)=3 (regal per chaos).
        ExchangeMarketSnapshot chaosRegalLeg = snapshot(chaos, regal, 1, 2, 1, 3, 10, 20, 5, 5);
        DivinationCardReward coveted = reward(9, "Regal Orb", 5);

        List<FlipOpportunity> opportunities =
                finder.find(List.of(chaosCardLeg, chaosRegalLeg), List.of(coveted), rate210);

        assertThat(opportunities).hasSize(1);
        FlipOpportunity opportunity = opportunities.get(0);
        double costChaos = 9.0 / 12;
        double resaleChaos = 5.0 / 3; // marketSellPrice (3), NOT a competitive/undercut price
        assertThat(opportunity.via().get(1).currency()).isEqualTo(regal);
        assertThat(opportunity.via().get(1).quantity()).isEqualTo(5.0);
        assertThat(opportunity.sell().get(0).quantity()).isCloseTo(resaleChaos, within(1e-6));
        assertThat(opportunity.marginPercent()).isCloseTo((resaleChaos - costChaos) / costChaos * 100, within(0.01));
    }

    @Test
    void find_rewardInDivine_convertsThroughTheAveragedReferenceRate() {
        ExchangeMarketSnapshot chaosCardLeg = snapshot(chaos, card, 1, 8, 1, 13, 100, 50, 1, 1);
        DivinationCardReward brothersGift = reward(1, "Divine Orb", 5);

        List<FlipOpportunity> opportunities = finder.find(List.of(chaosCardLeg), List.of(brothersGift), rate210);

        assertThat(opportunities).hasSize(1);
        FlipOpportunity opportunity = opportunities.get(0);
        assertThat(opportunity.via().get(1).currency()).isEqualTo(divine);
        assertThat(opportunity.via().get(1).quantity()).isEqualTo(5.0);
        assertThat(opportunity.sell().get(0).currency()).isEqualTo(chaos);
        assertThat(opportunity.sell().get(0).quantity()).isCloseTo(5.0 * 210.0, within(1e-6));
    }

    @Test
    void find_noChaosMarketForCard_fallsBackToDivineBuyLeg() {
        // Divine-card leg: priceAtLowest=1700, priceAtHighest=1900 -> buy=floor(1900)-1=1899, buyLegStock=highestStockA=80.
        ExchangeMarketSnapshot divineLeg = snapshot(divine, card, 1, 1700, 1, 1900, 30, 80, 1, 1);
        DivinationCardReward stackedDeckReward = reward(1, "Chaos Orb", 1);

        List<FlipOpportunity> opportunities = finder.find(List.of(divineLeg), List.of(stackedDeckReward), rate210);

        assertThat(opportunities).hasSize(1);
        FlipOpportunity opportunity = opportunities.get(0);
        assertThat(opportunity.start().get(0).currency()).isEqualTo(divine);
        assertThat(opportunity.start().get(0).quantity()).isCloseTo(1.0 / 1899, within(1e-9));
        assertThat(opportunity.volume()).isEqualTo(80);
    }

    @Test
    void find_noMarketForCardAgainstEitherBaseCurrency_isDropped() {
        DivinationCardReward stackedDeckReward = reward(8, "Chaos Orb", 5);

        assertThat(finder.find(List.of(), List.of(stackedDeckReward), rate210)).isEmpty();
    }

    @Test
    void find_rewardCurrencyHasNoChaosMarketOfItsOwn_isDropped() {
        ExchangeMarketSnapshot chaosCardLeg = snapshot(chaos, card, 1, 8, 1, 13, 100, 50, 1, 1);
        DivinationCardReward coveted = reward(9, "Regal Orb", 5); // no chaos<->regal snapshot supplied

        assertThat(finder.find(List.of(chaosCardLeg), List.of(coveted), rate210)).isEmpty();
    }

    @Test
    void find_nonPredictableCard_neverProducesAnOpportunity_evenWithPerfectMarketData() {
        ExchangeMarketSnapshot chaosCardLeg = snapshot(chaos, card, 1, 8, 1, 13, 100, 50, 1, 1);
        DivinationCardReward gambleCard = new DivinationCardReward(
                placeholder("Stacked Deck", Currency.ItemType.DIVINATION_CARD), 8, null, 0, false);

        assertThat(finder.find(List.of(chaosCardLeg), List.of(gambleCard), rate210)).isEmpty();
    }

    @Test
    void find_buyLegUndercutCollapsesBelowOne_isDroppedSafelyNoCrash() {
        // priceAtLowest=1, priceAtHighest=1.5 -> buy floors to 0 (invalid); no Divine leg to fall back to.
        ExchangeMarketSnapshot chaosCardLeg = snapshot(chaos, card, 2, 2, 2, 3, 40, 70, 1, 1);
        DivinationCardReward stackedDeckReward = reward(8, "Chaos Orb", 5);

        assertThat(finder.find(List.of(chaosCardLeg), List.of(stackedDeckReward), rate210)).isEmpty();
    }

    @Test
    void find_rewardCurrencyMarketSellPriceCollapsesToZero_isDroppedSafelyNoCrash() {
        // Same shape as the real Bulk Buy production bug: a wide in-hour spread
        // (0.5 vs 0.3) floors marketSellPrice to exactly 0, which would divide
        // rewardQuantity by zero if unguarded.
        ExchangeMarketSnapshot chaosCardLeg = snapshot(chaos, card, 1, 8, 1, 13, 100, 50, 1, 1);
        ExchangeMarketSnapshot chaosRegalLeg = snapshot(chaos, regal, 2, 1, 10, 3, 100, 50, 1, 1);
        DivinationCardReward coveted = reward(9, "Regal Orb", 5);

        assertThat(finder.find(List.of(chaosCardLeg, chaosRegalLeg), List.of(coveted), rate210)).isEmpty();
    }

    @Test
    void find_rateUnavailable_returnsEmptyRegardlessOfSnapshots() {
        ExchangeMarketSnapshot chaosCardLeg = snapshot(chaos, card, 1, 8, 1, 13, 100, 50, 1, 1);
        DivinationCardReward stackedDeckReward = reward(8, "Chaos Orb", 5);

        assertThat(finder.find(List.of(chaosCardLeg), List.of(stackedDeckReward), null)).isEmpty();
    }
}
